server/service/src/invoice_line/stock_in_line/insert/generate.rs

```diff
@@ -22,20 +24,28 @@ pub fn generate(
 ) -> Result<(Option<InvoiceRow>, InvoiceLineRow, Option<StockLineRow>), RepositoryError> {
     let store_preferences = get_store_preferences(connection, &existing_invoice_row.store_id)?;
 
-    let new_line = generate_line(input, item_row, existing_invoice_row.clone());
+    let new_line = generate_line(input.clone(), item_row, existing_invoice_row.clone());
 
     let mut new_line = match store_preferences.pack_to_one {
         true => convert_invoice_line_to_single_pack(new_line),
         false => new_line,
     };
 
-    let new_batch_option = if existing_invoice_row.status != InvoiceRowStatus::New {
+    let barcode_id = generate_barcode_id(&input, connection)?;
+
+    let new_batch_option = if should_upsert_batch(&input.r#type, &existing_invoice_row) {
         let new_batch = generate_batch(
-            &existing_invoice_row.store_id,
+            // If a stock line id is included in the input, use it
+            input.stock_line_id.is_some(),
```

Comment:
Not 100% happy with this... I wanted some level of idempotency for the request, so wanted to pass through the stock line ID from the input rather than generating a random one. However I'm using it here as though it is an "existing" stock line ID, and we should "update" it. It's an upsert, so it works fine, but it's a bit icky 🤷‍♀️